Mathlib/LinearAlgebra/RootSystem/Basic.lean

```diff
@@ -0,0 +1,352 @@
+`j`th root is the `k`th root, then the corresponding relationship holds for coroots. See also
+`RootSystem.coroot_eq_coreflection_of_root_eq`. -/
+lemma coroot_eq_coreflection_of_root_eq_of_span_eq_top [CharZero R] [NoZeroSMulDivisors R M]
+    (hsp : span R (range P.root) = ⊤)
+    {i j k : ι} (hk : P.root k = P.reflection i (P.root j)) :
+    P.coroot k = P.coreflection i (P.coroot j) :=
+  coroot_eq_coreflection_of_root_eq_of_span_eq_top' P.toPerfectPairing P.root P.coroot
+    P.coroot_root_two P.mapsTo_preReflection_root hsp hk
+
+end RootPairing
+
+namespace RootSystem
+
+open RootPairing
+
+variable {ι}
+variable (P : RootSystem ι R M N)
+
+/-- In characteristic zero if there is no torsion, a root system is determined entirely by its
+roots. -/
+@[ext]
+protected lemma ext [CharZero R] [NoZeroSMulDivisors R M]
+    {P₁ P₂ : RootSystem ι R M N}
+    (he : P₁.toLin = P₂.toLin)
+    (hr : P₁.root = P₂.root) :
+    P₁ = P₂ := by
+  suffices ∀ P₁ P₂ : RootSystem ι R M N, P₁.toLin = P₂.toLin → P₁.root = P₂.root →
+      range P₁.coroot ⊆ range P₂.coroot by
+    have h₁ := this P₁ P₂ he hr
+    have h₂ := this P₂ P₁ he.symm hr.symm
+    cases' P₁ with P₁
+    cases' P₂ with P₂
+    congr
+    exact RootPairing.ext he hr (le_antisymm h₁ h₂)
+  clear! P₁ P₂
+  rintro P₁ P₂ he hr - ⟨i, rfl⟩
+  use i
+  apply P₁.bijectiveRight.injective
+  apply Dual.eq_of_preReflection_mapsTo (P₁.ne_zero i) (finite_range P₁.root) P₁.span_eq_top
+  · exact hr ▸ he ▸ P₂.coroot_root_two i
+  · exact hr ▸ he ▸ P₂.mapsTo_preReflection_root i
+  · exact P₁.coroot_root_two i
+  · exact P₁.mapsTo_preReflection_root i
+
+/-- In characteristic zero if there is no torsion, to check that a family of roots form a root
+system, we do not need to check that the coroots are stable under reflections since this follows
+from the corresponding property for the roots. -/
+def mk' [CharZero R] [NoZeroSMulDivisors R M]
+    (p : PerfectPairing R M N)
+    (root : ι ↪ M)
+    (coroot : ι ↪ N)
+    (hp : ∀ i, p.toLin (root i) (coroot i) = 2)
+    (hs : ∀ i, MapsTo (preReflection (root i) (p.toLin.flip (coroot i))) (range root) (range root))
+    (hsp : span R (range root) = ⊤) :
+    RootSystem ι R M N where
+  toPerfectPairing := p
+  root := root
+  coroot := coroot
+  root_coroot_two := hp
+  mapsTo_preReflection_root := hs
+  span_eq_top := hsp
+  mapsTo_preReflection_coroot := by
+    rintro i - ⟨j, rfl⟩
+    obtain ⟨k, h⟩ := hs i (mem_range_self j)
+    exact ⟨k, coroot_eq_coreflection_of_root_eq_of_span_eq_top' p root coroot hp hs hsp h⟩
+
+/-- In characteristic zero if there is no torsion, if the `i`th reflection of the `j`th root is the
+`k`th root, then the corresponding relationship holds for coroots. -/
+lemma coroot_eq_coreflection_of_root_eq [CharZero R] [NoZeroSMulDivisors R M]
+    {i j k : ι} (hk : P.root k = P.reflection i (P.root j)) :
+    P.coroot k = P.coreflection i (P.coroot j) :=
+  P.coroot_eq_coreflection_of_root_eq_of_span_eq_top P.span_eq_top hk
+
```

Comment:
Can you please define `RootSystem.flip` as well?

Reply:
It is non-trivial that the coroots span if the roots span; it's fairly easy (but still some work) if you assume linearly-ordered scalars but quite a lot of work in general.

I definitely want this result but I think it should be a separate piece of work.

Reply:
Sure, agreed.